Ne plus réconcilier les usagers lors de la connexion FranceConnect

francois-ferrandis opened the pull request:
Closes #5429

# Contexte

Si vous êtes curieux, vous pouvez lire dans #5429 les raisons pour lesquelles FranceConnect interdit la réconciliation sur l'e-mail uniquement.

En résumé, dans le cadre d'une connexion par FranceConnect, c'est l'**identité de la personne** qui est le moyen de connexion. L'adresse e-mail fournie par FranceConnect n'est pas à considérer comme un moyen de connexion ni d'identification. Cette adresse e-mail peut être éventuellement utilisée pour envoyer des notifications.

# Solution

D'après [la doc de qualification FranceConnect](https://docs.partenaires.franceconnect.gouv.fr/fs/devenir-fs/projet-qualification-fs/) (point 17 de la doc) : 

> La réconciliation concerne un utilisateur qui aurait déjà un compte en local mais ne se serait jamais connecté au service via FranceConnect ou FranceConnect+.
>
> **Elle n’est pas obligatoire**, mais si elle est implémentée, elle doit respecter des conditions.

La réconciliation est donc un processus possible, et si il est désirable, il peut être mené à l'aide d'heuristiques basées obligatoirement sur l'identité pivot de la personne (nom de naissance, prénoms de naissance, date de naissance), et optionnellement sur d'autres données, selon le contexte.

**Le processus de réconciliation n'est donc à mettre en place que si le service proposé nécessite absolument d'éviter tout doublon dans le registre de ses usagers.** 

Dans le cas de notre application de prise de RDV, les usagers qui utilisent notre service vont très largement utiliser les liens des notifications pour aller consulter leur RDV. 

## Des chiffres

Sur RDV Solidarités :
- 1 126 351 usagers au total
- 28 529 usagers ont un sub FranceConnect
- 35 880 usagers on un mot de passe Devise
- 4 947 usagers on un sub FranceConnect **et** un mot de passe Devise

Sur RDV Service Public :
- 258 684 usagers au total
- 102 563 usagers ont un sub FranceConnect
- 33 150 usagers on un mot de passe Devise
- 4 162 usagers on un sub FranceConnect **et** un mot de passe Devise

# Migration

Nous pouvons donc déplacer la valeur de `email` vers `notification_email` pour tous les usagers qui se sont uniquement connectés par FranceConnect, et jamais par Devise. Cette migration concerne donc environ 23 000 `users` sur RDV Solidarités et 98 000 sur RDV Service Public. Peut-être à lancer hors période de travail, ça fait une grosse transaction.
 
# Changements fonctionnels

Nous avons donc désormais des usagers qui se loggent uniquement par FranceConnect, et qui n'ont donc pas de mail de login. Pour ces usagers nous cachons l'entrée de menu "Mon compte", et nous indiquons le `notification_email` dans un hint dans la section "Mes information" (screenshot ci-dessous).

## Usager qui a utilisé FranceConnect après le merge de cette PR (pas de mail de login Devise)

![image](https://github.com/user-attachments/assets/055157ac-b9d8-4856-8b6e-0e82aa52ef56)

## Usager sans FranceConnect, uniquement login Devise

![image](https://github.com/user-attachments/assets/8df7a1e4-30d5-40d0-ab2b-6fd4ecc4c7c0)


## Usager avec FranceConnect **et** login Devise

![image](https://github.com/user-attachments/assets/6aa07461-636f-47e9-8aae-1b8a7969f4a8)

francois-ferrandis committed:
Ne plus réconcilier les usagers lors de la connexion FranceConnect

francois-ferrandis committed:
Ne pas afficher le menu "Mon compte" que l'usager n'a pas d'email de login

francois-ferrandis committed:
Plutôt utiliser un champ disabled

francois-ferrandis committed:
Migre les usagers FC only de `email` vers `notification_email`

victormours reviewed (approved):
👍 vu ensemble, c'est effectivement un choix très raisonnable.

victormours committed:
Merge branch 'production' into frf/ne-plus-reconcilier-les-usagers-avec-franceconnect

adipasquale commented:
🙇 énorme merci francois-ferrandis (je pinge pas pour le respect des vacances 🏖️ ) pour le détail de l’issue et de la PR

ça me semble être un choix sensé pour avancer vers FC V2 sans trop se casser la tête et ça devrait assez largement nous économiser une classe de problèmes qu’en créer de nouveaux. 

je note quand même quelques petits sujets d’UX non-bloquants : 
- les usagers connectés via FC devraient quand même pouvoir supprimer leurs comptes RDVSP, ça se fait depuis la page mon compte
- le callout en dessous des champs nom prénom parle de champs qui n’apparaissent pas : nom de naissance, date et lieu de naissance. c’est un peu perturbant, on pourrait peut-être retirer la mention de ces champs invisibles ? peut-être que je rate qqch
- est-ce qu’on ne peut pas laisser l’usager changer son email de notif même s’il s’est connecté via FC ? 

très dispo pour échanger là-dessus @victormours

victormours committed:
Merge branch 'production' into frf/ne-plus-reconcilier-les-usagers-avec-franceconnect

victormours commented:
@adipasquale oui, je pense qu'il faut effectivement faire quelques modifs

> les usagers connectés via FC devraient quand même pouvoir supprimer leurs comptes RDVSP, ça se fait depuis la page mon compte

Tout à fait, je vais changer la logique de la vue pour permettre l'affichage de cette page, mais pas le changement de mot de passe.

> le callout en dessous des champs nom prénom parle de champs qui n’apparaissent pas : nom de naissance, date et lieu de naissance. c’est un peu perturbant, on pourrait peut-être retirer la mention de ces champs invisibles ? peut-être que je rate qqch

Oui, c'est étonnant, je vais voir si on peut corriger ça.

> est-ce qu’on ne peut pas laisser l’usager changer son email de notif même s’il s’est connecté via FC ?

Idéalement oui. C'est possible que ça soit plutôt une fonctionnalité supplémentaire qu'on ajoute plus tard (ça serait bien qu'on ai un peu de bande passante pourgérer les différents sujets autour des mails des usagers). Il y a aussi moyen que ça soit trivial à faire, je vais regarder si je peux embarquer ça dans cette pr.

victormours committed:
Update unit spec for new reconciliation logic

victormours committed:
Allow account deletion for users using SSO

victormours committed:
show Votre compte in layout

victormours committed:
Add logic to handle user having both devise and france connect

victormours committed:
update specs

victormours committed:
uniform naming

victormours committed:
simplification du message franceconnect frozen fields
La liste des champs qui sont affichés dépend d'une logique complexe, qu'on ne va pas reproduire dans le messag

victormours commented:
J'ai fait quelques modifs pour qu'on permette aux usagers qui utilisent FranceConnect et la connexion par email/mot de passe de continuer à pouvoir faire les deux.

victormours committed:
Fix typo

adipasquale reviewed (approved):
👍 merci @victormours pour tout ce boulot

J’ai relu assez attentivement tes nouveaux commits mais je n’ai pas re-testé. Ça me paraît tout bon je n’ai rien à rajouter je pense qu’on peut déployer ça 🚀

victormours committed:
Merge branch 'production' into frf/ne-plus-reconcilier-les-usagers-avec-franceconnect
